posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+                    self.current_file_index,
+                    part_number,
+                    self.upload_id,
+                    part_size_mb,
+                    upload_time,
+                    upload_speed_mbps,
+                )
+                return part_info
+
+        except Exception:
+            await self.logger.aexception(
+                "Failed to upload file number %s part %s with upload id %s",
+                self.current_file_index,
+                part_number,
+                self.upload_id,
+            )
+            raise
+        finally:
+            # Always release memory semaphore
+            self.memory_semaphore.release()
+            # Explicitly delete data to free memory immediately
+            del data
+
+    def _get_current_key(self) -> str:
+        """Generate the key for the current file"""
+        key_prefix = get_s3_key_prefix(self.s3_inputs)
+
+        try:
+            file_extension = FILE_FORMAT_EXTENSIONS[self.s3_inputs.file_format]
+        except KeyError:
+            raise UnsupportedFileFormatError(self.s3_inputs.file_format, "S3")
+
+        base_file_name = f"{self.s3_inputs.data_interval_start}-{self.s3_inputs.data_interval_end}"
+        # to maintain backwards compatibility with the old file naming scheme
+        if self.s3_inputs.max_file_size_mb is not None:
+            base_file_name = f"{base_file_name}-{self.current_file_index}"
+        if self.s3_inputs.compression is not None:
+            file_name = base_file_name + f".{file_extension}.{COMPRESSION_EXTENSIONS[self.s3_inputs.compression]}"
+        else:
+            file_name = base_file_name + f".{file_extension}"
+
+        key = posixpath.join(key_prefix, file_name)
+
+        if posixpath.isabs(key):
+            # Keys are relative to root dir, so this would add an extra "/"
+            key = posixpath.relpath(key, "/")
+
+        return key
+
+    async def _start_new_file(self):
+        """Start a new file (reset state for file splitting)"""
+        self.current_file_index += 1
+        self.current_file_size = 0
+        self.part_counter = 1
+        self.upload_id = None
+        self.pending_uploads.clear()
+        self.completed_parts.clear()
+        await self.logger.adebug("Starting new file: %s", self._get_current_key())
+
+    async def _finalize_current_file(self):
+        """Finalize the current file before starting a new one"""
+        if self.current_file_size == 0:
+            return  # Nothing to finalize
+
+        try:
+            # Upload any remaining data in buffer
+            if len(self.current_buffer) > 0:
+                # for now, let's ignore this so our tests use multi part uploads too
+                # if not self.upload_id:
+                #     # Small file - direct upload
+                #     await self._single_file_upload_current()
+                # else:
+                # Upload final part
```

Comment:
Do we need this uncommented for production? Could we have a 
```
if not settings.TEST and not settings.DEBUG:
...
```
so that it works both locally and in production?

Reply:
could do, although then I suppose you're not testing the code that runs in production